File: graphColoring_branchAndBound.cpp

#include <iostream>
#include <vector>
#include <algorithm>
#include <fstream>
#include <chrono>
#include <climits>

using namespace std;

const int MAXN = 1000; // Maximum number of vertices

int n; // Number of vertices
vector<int> adj[MAXN]; // Adjacency list of the graph
int color[MAXN]; // Array of vertex colors
int minColors; // Minimum number of colors found so far

bool isSafe(int v, int c) {
    for (int u : adj[v]) {
        if (color[u] == c) {
            return false;
        }
    }
    return true;
}

void graphColoringUtil(int v, int colorsUsed) {
    if (v == n) {
        minColors = min(minColors, colorsUsed);
        return;
    }

    for (int c = 1; c <= colorsUsed + 1; c++) {
        if (isSafe(v, c)) {
            color[v] = c;
            graphColoringUtil(v + 1, colorsUsed);
            color[v] = 0;
        }
    }
}

int graphColoring() {
    minColors = INT_MAX;
    graphColoringUtil(0, 0);
    return minColors;
}

int main() {
    ifstream fin("graphcoloring.in");
    ofstream fout("graphcoloring.out");

    // Read input
    int m; // Number of edges
    fin >> n >> m;

    for (int i = 0; i < m; i++) {
        int u, v;
        fin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    auto start = chrono::steady_clock::now();

    int numColors = graphColoring();

    auto end = chrono::steady_clock::now();
    double elapsedTime = chrono::duration_cast<chrono::microseconds>(end - start).count();

    // Print colors
    for (int i = 0; i < n; i++) {
        fout << "Vertex " << i + 1 << " is colored " << color[i] << endl;
        cout << "Vertex " << i + 1 << " is colored " << color[i] << endl;
    }

    fout << "Number of colors needed: " << numColors << endl;
    cout << "Number of colors needed: " << numColors << endl;

    fout << "Elapsed time (microseconds): " << elapsedTime << endl;
    cout << "Elapsed time (microseconds): " << elapsedTime << endl;

    return 0;
}

File: graphColoring_BranchAndBound.py
import sys
import time

def is_safe(v, c, colors):
    for u in adj[v]:
        if u < v and colors[u] == c:
            return False, colors[u]
    return True, colors[v]

def graph_coloring_util(v, colors):
    if v == n:
        max_color = max(colors)
        min_colors[0] = min(min_colors[0], max_color + 1)
        return

    for c in range(n + 1):
        safe, assigned_color = is_safe(v, c, colors)
        if safe:
            colors[v] = c
            graph_coloring_util(v + 1, colors)
            return assigned_color
    
    colors[v] = 0

def graph_coloring():
    min_colors[0] = sys.maxsize
    start_time = time.time()
    graph_coloring_util(0, colors)
    end_time = time.time()
    return min_colors[0], end_time - start_time

if __name__ == "__main__":
    # Read input from file
    with open("graphcoloring1000.in", "r") as fin:
        n, m = map(int, fin.readline().split())
        adj = [[] for _ in range(n)]
        colors = [0] * n
        min_colors = [0]

        for _ in range(m):
            u, v = map(int, fin.readline().split())
            adj[u - 1].append(v - 1)
            adj[v - 1].append(u - 1)

    # Solve graph coloring problem and measure execution time
    num_colors, execution_time = graph_coloring()

    # Store vertex colors
    vertex_colors = list(colors)

    # Write output to file
    with open("graphcoloring1000.out", "w") as fout:
        # Print colors
        for i in range(n):
            fout.write(f"Vertex {i + 1} is colored {vertex_colors[i]}\n")

        fout.write(f"Number of colors needed: {num_colors}\n")
        fout.write(f"Execution time: {execution_time} seconds\n")
